File: src/main/java/de/hanno/hpengine/util/gui/MaterialView.java
package de.hanno.hpengine.util.gui;

import com.alee.extended.panel.GridPanel;
import com.alee.extended.panel.GroupPanel;
import com.alee.extended.panel.WebComponentPanel;
import com.alee.laf.button.WebButton;
import com.alee.laf.combobox.WebComboBox;
import com.alee.laf.label.WebLabel;
import com.alee.laf.panel.WebPanel;
import com.alee.laf.scroll.WebScrollPane;
import com.alee.laf.slider.WebSlider;
import com.alee.laf.text.WebTextField;
import com.alee.managers.notification.NotificationIcon;
import com.alee.managers.notification.NotificationManager;
import com.alee.managers.notification.WebNotificationPopup;
import de.hanno.hpengine.engine.Engine;
import de.hanno.hpengine.event.MaterialChangedEvent;
import de.hanno.hpengine.renderer.GraphicsContext;
import de.hanno.hpengine.renderer.command.GetMaterialCommand;
import de.hanno.hpengine.renderer.command.InitMaterialCommand;
import de.hanno.hpengine.renderer.material.Material;
import de.hanno.hpengine.renderer.material.Material.MAP;
import de.hanno.hpengine.renderer.material.MaterialInfo;
import de.hanno.hpengine.texture.Texture;
import de.hanno.hpengine.texture.TextureFactory;
import de.hanno.hpengine.util.gui.input.*;
import org.apache.commons.io.FileUtils;
import org.lwjgl.util.vector.Vector3f;

import java.awt.*;
import java.io.File;
import java.io.IOException;
import java.util.*;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;

public class MaterialView extends WebPanel {

	private Material material;
	private WebTextField nameField;

	public MaterialView(Material material) {
		setUndecorated(true);
		this.setSize(600, 600);
		setMargin(20);
		
		init(material);
	}

	private void init(Material material) {
		this.material = material;
		this.removeAll();
		List<Component> panels = new ArrayList<>();
		
		addTexturePanel(panels);
        addValuePanels(panels);

        WebButton saveButton = new WebButton("Save");
        saveButton.addActionListener(e -> {
        	Material toSave = null;
        	if(!nameField.getText().equals(material.getMaterialInfo().name)) {
        		MaterialInfo newInfo = new MaterialInfo(material.getMaterialInfo()).setName(nameField.getText());
				CompletableFuture<InitMaterialCommand.MaterialResult> future = GraphicsContext.getInstance().execute(() -> {
					return new GetMaterialCommand(newInfo).execute(Engine.getInstance());
				});
				InitMaterialCommand.MaterialResult result;
				try {
        			result = future.get(1, TimeUnit.MINUTES);
					if(result.material != null) {
                        toSave = result.material;
						showNotification(NotificationIcon.plus, "Material changed");
					} else {
						showNotification(NotificationIcon.error, "Not able to change material");
					}
        		} catch (Exception e1) {
        			e1.printStackTrace();
        		}

        	} else {
        		toSave = material;
        		Material.write(toSave, toSave.getMaterialInfo().name);
        	}

        	addMaterialInitCommand(toSave);
        });
        
        Component[] components = new Component[panels.size()];
        panels.toArray(components);

        WebScrollPane materialAttributesPane = new WebScrollPane(new GridPanel ( panels.size(), 1, components));
        this.setLayout(new BorderLayout());
        this.add(materialAttributesPane, BorderLayout.CENTER);
        this.add(saveButton, BorderLayout.SOUTH);
	}

	private void addTexturePanel(List<Component> panels) {
		
		WebComponentPanel webComponentPanel = new WebComponentPanel ( true );
        webComponentPanel.setElementMargin ( 4 );

        addExistingTexturesPanels(webComponentPanel);
        addMissingTexturesPanels(webComponentPanel);
        
        panels.add(webComponentPanel);
	}

	private void addMissingTexturesPanels(WebComponentPanel webComponentPanel) {
		EnumSet<MAP> missingMaps = EnumSet.allOf(MAP.class);
		missingMaps.removeAll(material.getMaterialInfo().maps.getTextures().keySet());
		for (MAP map : missingMaps) {
			WebLabel label = new WebLabel ( map.name() );
	        
	        List<Texture> textures = getAllTexturesSorted();
	        WebComboBox select = new WebComboBox(new Vector<Texture>(textures));
	        select.setSelectedIndex(-1);
	        
	        select.addActionListener(e -> {
	        	WebComboBox cb = (WebComboBox) e.getSource();
	        	Texture selectedTexture = textures.get(cb.getSelectedIndex());
				if(selectedTexture == null) { return; }
	        	material.getMaterialInfo().maps.put(map, selectedTexture);
	        	addMaterialInitCommand(material);
	        });
	        
	        WebButton removeTextureButton = new WebButton("Remove");
	        removeTextureButton.addActionListener(e -> {
	        	material.getMaterialInfo().maps.getTextures().remove(map);
		        select.setSelectedIndex(-1);
	        	addMaterialInitCommand(material);
	        });

            GroupPanel groupPanel = new GroupPanel(4, label, select, removeTextureButton);
			webComponentPanel.addElement(groupPanel);
		}
	}

	private void addExistingTexturesPanels(WebComponentPanel webComponentPanel) {
		for (MAP map : material.getMaterialInfo().maps.getTextures().keySet()) {
			Texture texture = material.getMaterialInfo().maps.getTextures().get(map);
			
	        WebLabel label = new WebLabel ( map.name() );
	        
	        List<Texture> textures = getAllTexturesSorted();
	        WebComboBox select = new WebComboBox(new Vector<Texture>(textures));
	        
	        int assignedTexture = textures.indexOf(texture);
	        select.setSelectedIndex(assignedTexture);
	        
	        select.addActionListener(e -> {
	        	WebComboBox cb = (WebComboBox) e.getSource();
	        	Texture selectedTexture = textures.get(cb.getSelectedIndex());
	        	material.getMaterialInfo().maps.put(map, selectedTexture);
	        	Engine.getEventBus().post(new MaterialChangedEvent(material));
	        });
	        
	        WebButton removeTextureButton = new WebButton("Remove");
	        removeTextureButton.addActionListener(e -> {
	        	material.getMaterialInfo().maps.getTextures().remove(map);
	        	material.getMaterialInfo().maps.getTextureNames().remove(map);
	        	addMaterialInitCommand(material);
	        });
	        
	        GroupPanel groupPanel = new GroupPanel ( 4, label, select, removeTextureButton );
			webComponentPanel.addElement(groupPanel);
		}
	}

	private void addValuePanels(List<Component> panels) {
		WebComponentPanel webComponentPanel = new WebComponentPanel ( true );
        webComponentPanel.setElementMargin ( 4 );
        webComponentPanel.setLayout(new FlowLayout());

        nameField = new WebTextField(material.getName());
		webComponentPanel.addElement(nameField);
        
        webComponentPanel.addElement(new WebFormattedVec3Field("Diffuse", material.getDiffuse()) {
			@Override
			public void onValueChange(Vector3f current) {
				material.setDiffuse(current);
	        	Engine.getEventBus().post(new MaterialChangedEvent(material));
			}
		});
        webComponentPanel.addElement(new ColorChooserButton("Diffuse", new ColorChooserFrame() {
			@Override
			public void onColorChange(Vector3f color) {
				material.setDiffuse(color);
	        	Engine.getEventBus().post(new MaterialChangedEvent(material));
			}
		}));

//        webComponentPanel.addElement(new WebFormattedVec3Field("SpecularColor", material.getSpecular()) {
//			@Override
//			public void onValueChange(Vector3f current) {
//				material.setSpecular(current);
//			}
//		});
//        webComponentPanel.addElement(new ColorChooserButton("SpecularColor", new ColorChooserFrame() {
//			@Override
//			public void onColorChange(Vector3f color) {
//				material.setSpecular(color);
//			}
//		}));
//        {
//	        LimitedWebFormattedTextField specularExponentInput = new LimitedWebFormattedTextField(1, 10000) {
//				@Override
//				public void onChange(float currentValue) {
//					material.setSpecularCoefficient(currentValue);
//				}
//			};
//			specularExponentInput.setValue(material.getSpecularCoefficient());
//	        GroupPanel groupPanel = new GroupPanel ( 4, new WebLabel("Specular Power"), specularExponentInput );
//	        TooltipManager.setTooltip ( groupPanel, "1 soft, 100 hard highlights", TooltipWay.up );
//	        webComponentPanel.addElement(groupPanel);
//        }
        {
            LimitedWebFormattedTextField roughnessInput = new LimitedWebFormattedTextField(0, 1) {
    			@Override
    			public void onChange(float currentValue) {
    				material.setRoughness(currentValue);
    			}
    		};
    		roughnessInput.setValue(material.getRoughness());
            {
            	SliderInput roughnessSliderInput = new SliderInput("Roughness", WebSlider.HORIZONTAL, 0, 100, (int)(material.getRoughness()*100)) {
    				
    				@Override
    				public void onValueChange(int value, int delta) {
    					roughnessInput.setValue(((float)value/100f));
    					material.setRoughness(((float)value/100f));
    		        	Engine.getEventBus().post(new MaterialChangedEvent(material));
    				}
    			};
                GroupPanel groupPanelRoughness = new GroupPanel ( 4, new WebLabel("Roughness"), roughnessInput, roughnessSliderInput );
                webComponentPanel.addElement(groupPanelRoughness);
            }
        }
        {
            LimitedWebFormattedTextField metallicInput = new LimitedWebFormattedTextField(0, 1) {
    			@Override
    			public void onChange(float currentValue) {
    				material.setMetallic(currentValue);
    			}
    		};
    		metallicInput.setValue(material.getMetallic());
            {
            	SliderInput metallicSliderInput = new SliderInput("Metallic", WebSlider.HORIZONTAL, 0, 100, (int)(material.getMetallic()*100)) {
    				
    				@Override
    				public void onValueChange(int value, int delta) {
    					metallicInput.setValue(((float)value/100f));
    					material.setMetallic(((float)value/100f));
    		        	Engine.getEventBus().post(new MaterialChangedEvent(material));
    				}
    			};
    			
                GroupPanel groupPanelMetallic = new GroupPanel ( 4, new WebLabel("Metallic"), metallicInput, metallicSliderInput );
                webComponentPanel.addElement(groupPanelMetallic);
            }
        }
        {
            LimitedWebFormattedTextField ambientInput = new LimitedWebFormattedTextField(0, 1) {
    			@Override
    			public void onChange(float currentValue) {
    				material.setAmbient(currentValue);
    			}
    		};
    		ambientInput.setValue(material.getAmbient());
            {
            	SliderInput ambientSliderInput = new SliderInput("Ambient/Emmissive", WebSlider.HORIZONTAL, 0, 100, (int)(material.getAmbient()*100)) {
    				
    				@Override
    				public void onValueChange(int value, int delta) {
    					ambientInput.setValue(((float)value/100f));
    					material.setAmbient(((float)value/100f));
    		        	Engine.getEventBus().post(new MaterialChangedEvent(material));
    				}
    			};
    			
                GroupPanel groupPanelAmbient = new GroupPanel(4, new WebLabel("Ambient"), ambientInput, ambientSliderInput);
                webComponentPanel.addElement(groupPanelAmbient);
            }
        }
        {
            LimitedWebFormattedTextField transparencyInput = new LimitedWebFormattedTextField(0, 1) {
    			@Override
    			public void onChange(float currentValue) {
    				material.setTransparency(currentValue);
    			}
    		};
    		transparencyInput.setValue(material.getTransparency());
            {
            	SliderInput transparencySliderInput = new SliderInput("Transparency", WebSlider.HORIZONTAL, 0, 100, (int)(material.getTransparency()*100)) {
    				
    				@Override
    				public void onValueChange(int value, int delta) {
    					transparencyInput.setValue(((float)value/100f));
    					material.setTransparency(((float)value/100f));
    		        	Engine.getInstance().getEventBus().post(new MaterialChangedEvent(material));
    				}
    			};
    			
                GroupPanel groupPanelTransparency = new GroupPanel(4, new WebLabel("Transparency"), transparencyInput, transparencySliderInput);
                webComponentPanel.addElement(groupPanelTransparency);
            }
        }
        {
            LimitedWebFormattedTextField parallaxScaleInput = new LimitedWebFormattedTextField(0, 1) {
    			@Override
    			public void onChange(float currentValue) {
    				material.setParallaxScale(currentValue);
    			}
    		};
    		parallaxScaleInput.setValue(material.getParallaxScale());
            {
            	SliderInput parallaxScaleSliderInput = new SliderInput("Parallax Scale", WebSlider.HORIZONTAL, 0, 100, (int)(material.getParallaxScale()*100)) {
    				
    				@Override
    				public void onValueChange(int value, int delta) {
    					parallaxScaleInput.setValue(((float)value/100f));
    					material.setParallaxScale(((float)value/100f));
                        Engine.getInstance().getEventBus().post(new MaterialChangedEvent(material));
    				}
    			};
    			
                GroupPanel groupPanelParallaxScale = new GroupPanel(4, new WebLabel("Parallax Scale"), parallaxScaleInput, parallaxScaleSliderInput);
                webComponentPanel.addElement(groupPanelParallaxScale);
            }
        }
        {
            LimitedWebFormattedTextField parallaxBiasInput = new LimitedWebFormattedTextField(0, 1) {
    			@Override
    			public void onChange(float currentValue) {
    				material.setParallaxBias(currentValue);
    			}
    		};
    		parallaxBiasInput.setValue(material.getParallaxBias());
            {
            	SliderInput parallaxBiasSliderInput = new SliderInput("Parallax Bias", WebSlider.HORIZONTAL, 0, 100, (int)(material.getParallaxScale()*100)) {
    				
    				@Override
    				public void onValueChange(int value, int delta) {
    					parallaxBiasInput.setValue(((float)value/100f));
    					material.setParallaxBias(((float)value/100f));
                        Engine.getInstance().getEventBus().post(new MaterialChangedEvent(material));
    				}
    			};
    			
                GroupPanel groupPanelParallaxBias = new GroupPanel(4, new WebLabel("Parallax Bias"), parallaxBiasInput, parallaxBiasSliderInput);
                webComponentPanel.addElement(groupPanelParallaxBias);
            }
        }

		{
			WebComboBox materialTypeInput = new WebComboBox((EnumSet.allOf(Material.MaterialType.class)).toArray());
			materialTypeInput.addActionListener(e -> {
				Material.MaterialType selected = (Material.MaterialType) materialTypeInput.getSelectedItem();
				material.setMaterialType(selected);
                Engine.getInstance().getEventBus().post(new MaterialChangedEvent(material));
			});
			materialTypeInput.setSelectedItem(material.getMaterialType());
			GroupPanel materialTypePanel = new GroupPanel(4, new WebLabel("Maeterial Type"), materialTypeInput);
			webComponentPanel.addElement(materialTypePanel);
		}
		{
			WebComboBox environmentMapInput = new WebComboBox((EnumSet.allOf(Material.ENVIRONMENTMAPTYPE.class)).toArray());
			environmentMapInput.addActionListener(e -> {
				Material.ENVIRONMENTMAPTYPE selected = (Material.ENVIRONMENTMAPTYPE) environmentMapInput.getSelectedItem();
				material.setEnvironmentMapType(selected);
			});
			environmentMapInput.setSelectedItem(material.getEnvironmentMapType());
			GroupPanel groupPanelEnironmentMapType = new GroupPanel ( 4, new WebLabel("Environment map type"), environmentMapInput );
			webComponentPanel.addElement(groupPanelEnironmentMapType);
		}
		
		panels.add(webComponentPanel);
	}

	private void copyShaderIfNotPresent(File chosenFile, File shaderFileInWorkDir) {
		if (!shaderFileInWorkDir.exists()) {
			try {
				FileUtils.copyFile(chosenFile, shaderFileInWorkDir);
			} catch (IOException e) {
				e.printStackTrace();
			}
		}
	}
	
	private void addMaterialInitCommand(Material material) {
		CompletableFuture<InitMaterialCommand.MaterialResult> future = GraphicsContext.getInstance().execute(() -> {
			return new InitMaterialCommand(material).execute(Engine.getInstance());
		});

		InitMaterialCommand.MaterialResult result;
		try {
			result = future.get(1, TimeUnit.MINUTES);
			if(result.isSuccessful()) {
				showNotification(NotificationIcon.plus, "Material changed");

				init(result.material);
				Engine.getEventBus().post(new MaterialChangedEvent(material));
			} else {
				showNotification(NotificationIcon.error, "Not able to change material");
			}
		} catch (Exception e1) {
			e1.printStackTrace();
			showNotification(NotificationIcon.error, "Not able to change material");
		}
	}

	private List<Texture> getAllTexturesSorted() {
        List<Texture> temp = (List<Texture>) TextureFactory.getInstance().TEXTURES.values().stream().sorted(new Comparator<Texture>() {
			@Override
			public int compare(Texture o1, Texture o2) {
				return (o1.getPath().compareTo(o2.getPath()));
			}
		}).collect(Collectors.toList());
        return temp;
	}
	private void showNotification(NotificationIcon icon, String text) {
		final WebNotificationPopup notificationPopup = new WebNotificationPopup();
		notificationPopup.setIcon(icon);
		notificationPopup.setDisplayTime( 2000 );
		notificationPopup.setContent(new WebLabel(text));
		NotificationManager.showNotification(notificationPopup);
	}
}

File: engine/src/main/java/de/hanno/hpengine/util/gui/MaterialView.kt
package de.hanno.hpengine.util.gui

import com.alee.extended.panel.GridPanel
import com.alee.extended.panel.GroupPanel
import com.alee.extended.panel.WebComponentPanel
import com.alee.laf.button.WebButton
import com.alee.laf.combobox.WebComboBox
import com.alee.laf.label.WebLabel
import com.alee.laf.panel.WebPanel
import com.alee.laf.scroll.WebScrollPane
import com.alee.laf.slider.WebSlider
import com.alee.laf.text.WebTextField
import com.alee.managers.notification.NotificationIcon
import com.alee.managers.notification.NotificationManager
import com.alee.managers.notification.WebNotificationPopup
import de.hanno.hpengine.engine.Engine
import de.hanno.hpengine.engine.event.MaterialChangedEvent
import de.hanno.hpengine.engine.graphics.renderer.command.GetMaterialCommand
import de.hanno.hpengine.engine.graphics.renderer.command.InitMaterialCommand
import de.hanno.hpengine.engine.model.material.SimpleMaterial
import de.hanno.hpengine.engine.model.material.SimpleMaterial.MAP
import de.hanno.hpengine.engine.model.material.SimpleMaterialInfo
import de.hanno.hpengine.engine.model.texture.Texture
import de.hanno.hpengine.engine.model.texture.TextureDimension2D
import de.hanno.hpengine.util.gui.input.*
import org.apache.commons.io.FileUtils
import org.joml.Vector3f
import java.awt.BorderLayout
import java.awt.Component
import java.awt.FlowLayout
import java.io.File
import java.io.IOException
import java.util.*
import kotlin.Comparator

class MaterialView(private val engine: Engine<*>, var material: SimpleMaterial) : WebPanel() {
    private val nameField = WebTextField(material.materialInfo.name)

    private val allTexturesSorted: List<Texture<*>>
        get() = engine.textureManager.textures.values.sortedWith(Comparator { o1, o2 -> o1.toString().compareTo(o2.toString()) })

    init {
        isUndecorated = true
        this.setSize(600, 600)
        setMargin(5)

        init(material)
    }

    private fun init(material: SimpleMaterial) {
        this.material = material
        nameField.text = material.materialInfo.name
        this.removeAll()
        val panels = ArrayList<Component>()

        addTexturePanel(panels)
        addValuePanels(panels)

        val components = panels.toTypedArray()

        val materialAttributesPane = WebScrollPane(GridPanel(panels.size, 1, *components))
        this.layout = BorderLayout()
        this.add(materialAttributesPane, BorderLayout.CENTER)
    }

    private fun addTexturePanel(panels: MutableList<Component>) {

        val webComponentPanel = WebComponentPanel(true)
        webComponentPanel.setElementMargin(4)

        addExistingTexturesPanels(webComponentPanel)
        addMissingTexturesPanels(webComponentPanel)

        panels.add(webComponentPanel)
    }

    private fun addMissingTexturesPanels(webComponentPanel: WebComponentPanel) {
        val missingMaps = EnumSet.allOf(MAP::class.java)
        missingMaps.removeAll(material.materialInfo.maps.keys)
        for (map in missingMaps) {
            val label = WebLabel(map.name)

            val textures = allTexturesSorted
            val select = WebComboBox(Vector(textures))
            select.selectedIndex = -1

            select.addActionListener { e ->
                val cb = e.source as WebComboBox
                val selectedTexture = textures[cb.selectedIndex]
                engine.scene.materialManager.changeMaterial(material.put(map, selectedTexture as Texture<TextureDimension2D>))
//                material.materialInfo.maps.put(map, selectedTexture)
                addMaterialInitCommand(material)
            }

            val removeTextureButton = WebButton("Remove")
            removeTextureButton.addActionListener { e ->
                engine.scene.materialManager.changeMaterial(material.remove(map))
//                material.materialInfo.maps.remove(map)
                select.selectedIndex = -1
                addMaterialInitCommand(material)
            }

            val groupPanel = GroupPanel(4, label, select, removeTextureButton)
            webComponentPanel.addElement(groupPanel)
        }
    }

    private fun addExistingTexturesPanels(webComponentPanel: WebComponentPanel) {
        for (map in material.materialInfo.maps.keys) {
            val texture: Texture<TextureDimension2D> = material.materialInfo.maps[map]!!

            val label = WebLabel(map.name)

            val textures = allTexturesSorted
            val select = WebComboBox(Vector(textures))

            val assignedTexture = textures.indexOf(texture)
            select.selectedIndex = assignedTexture

            select.addActionListener { e ->
                val cb = e.source as WebComboBox
                val selectedTexture: Texture<TextureDimension2D> = textures[cb.selectedIndex] as Texture<TextureDimension2D>
                engine.scene.materialManager.changeMaterial(material.materialInfo.put(map, selectedTexture))
//                material.materialInfo.maps.put(map, selectedTexture)
                engine.eventBus.post(MaterialChangedEvent(material))
            }

            val removeTextureButton = WebButton("Remove")
            removeTextureButton.addActionListener { e ->
                engine.scene.materialManager.changeMaterial(material.materialInfo.remove(map))
//                material.materialInfo.maps.remove(map)
                addMaterialInitCommand(material)
            }

            val groupPanel = GroupPanel(4, label, select, removeTextureButton)
            webComponentPanel.addElement(groupPanel)
        }
    }

    private fun addValuePanels(panels: MutableList<Component>) {
        val webComponentPanel = WebComponentPanel(true)
        webComponentPanel.setElementMargin(4)
        webComponentPanel.layout = FlowLayout()

        webComponentPanel.addElement(nameField)

        webComponentPanel.addElement(object : WebFormattedVec3Field("Diffuse", material.materialInfo.diffuse) {
            override fun onValueChange(current: Vector3f) {
                engine.scene.materialManager.changeMaterial(material.materialInfo.copyXXX(diffuse = current))
                engine.eventBus.post(MaterialChangedEvent(material))
            }
        })
        webComponentPanel.addElement(ColorChooserButton("Diffuse", object : ColorChooserFrame() {
            override fun onColorChange(color: Vector3f) {
                engine.scene.materialManager.changeMaterial(material.materialInfo.copyXXX(diffuse = color))
                engine.eventBus.post(MaterialChangedEvent(material))
            }
        }))

        run {
            val roughnessInput = object : LimitedWebFormattedTextField(0f, 1f) {
                override fun onChange(currentValue: Float) {
                    engine.scene.materialManager.changeMaterial(material.materialInfo.copyXXX(roughness = currentValue))
                    engine.eventBus.post(MaterialChangedEvent(material))
                }
            }
            roughnessInput.setValue(material.materialInfo.roughness)
            run {
                val roughnessSliderInput = object : SliderInput("Roughness", WebSlider.HORIZONTAL, 0, 100, (material.materialInfo.roughness * 100).toInt()) {

                    override fun onValueChange(value: Int, delta: Int) {
                        roughnessInput.setValue(value.toFloat() / 100f)
                        engine.scene.materialManager.changeMaterial(material.materialInfo.copyXXX(roughness = value.toFloat() / 100f))
                        engine.eventBus.post(MaterialChangedEvent(material))
                    }
                }
                val groupPanelRoughness = GroupPanel(4, WebLabel("Roughness"), roughnessInput, roughnessSliderInput)
                webComponentPanel.addElement(groupPanelRoughness)
            }
        }
        run {
            val metallicInput = object : LimitedWebFormattedTextField(0f, 1f) {
                override fun onChange(currentValue: Float) {
                    engine.scene.materialManager.changeMaterial(material.materialInfo.copyXXX(metallic = currentValue))
                    engine.eventBus.post(MaterialChangedEvent(material))
                }
            }
            metallicInput.setValue(material.materialInfo.metallic)
            run {
                val metallicSliderInput = object : SliderInput("Metallic", WebSlider.HORIZONTAL, 0, 100, (material.materialInfo.metallic * 100).toInt()) {

                    override fun onValueChange(value: Int, delta: Int) {
                        metallicInput.setValue(value.toFloat() / 100f)
                        engine.scene.materialManager.changeMaterial(material.materialInfo.copyXXX(metallic = value.toFloat() / 100f))
                        engine.eventBus.post(MaterialChangedEvent(material))
                    }
                }

                val groupPanelMetallic = GroupPanel(4, WebLabel("Metallic"), metallicInput, metallicSliderInput)
                webComponentPanel.addElement(groupPanelMetallic)
            }
        }
        run {
            val ambientInput = object : LimitedWebFormattedTextField(0f, 1f) {
                override fun onChange(currentValue: Float) {
                    engine.scene.materialManager.changeMaterial(material.materialInfo.copyXXX(ambient = currentValue))
                    engine.eventBus.post(MaterialChangedEvent(material))
                }
            }
            ambientInput.setValue(material.materialInfo.ambient)
            run {
                val ambientSliderInput = object : SliderInput("Ambient/Emmissive", WebSlider.HORIZONTAL, 0, 100, (material.materialInfo.ambient * 100).toInt()) {

                    override fun onValueChange(value: Int, delta: Int) {
                        ambientInput.setValue(value.toFloat() / 100f)
                        engine.scene.materialManager.changeMaterial(material.materialInfo.copyXXX(ambient = value.toFloat() / 100f))
                        engine.eventBus.post(MaterialChangedEvent(material))
                    }
                }

                val groupPanelAmbient = GroupPanel(4, WebLabel("Ambient"), ambientInput, ambientSliderInput)
                webComponentPanel.addElement(groupPanelAmbient)
            }
        }
        run {
            val transparencyInput = object : LimitedWebFormattedTextField(0f, 1f) {
                override fun onChange(currentValue: Float) {
                    engine.scene.materialManager.changeMaterial(material.materialInfo.copyXXX(transparency = currentValue))
                    engine.eventBus.post(MaterialChangedEvent(material))
                }
            }
            transparencyInput.setValue(material.materialInfo.transparency)
            run {
                val transparencySliderInput = object : SliderInput("Transparency", WebSlider.HORIZONTAL, 0, 100, (material.materialInfo.transparency * 100).toInt()) {

                    override fun onValueChange(value: Int, delta: Int) {
                        transparencyInput.setValue(value.toFloat() / 100f)
                        engine.scene.materialManager.changeMaterial(material.materialInfo.copyXXX(transparency = value.toFloat() / 100f))
                        engine.eventBus.post(MaterialChangedEvent(material))
                    }
                }

                val groupPanelTransparency = GroupPanel(4, WebLabel("Transparency"), transparencyInput, transparencySliderInput)
                webComponentPanel.addElement(groupPanelTransparency)
            }
        }
        run {
            val parallaxScaleInput = object : LimitedWebFormattedTextField(0f, 1f) {
                override fun onChange(currentValue: Float) {
                    engine.scene.materialManager.changeMaterial(material.materialInfo.copyXXX(parallaxScale = currentValue))
                    engine.eventBus.post(MaterialChangedEvent(material))
                }
            }
            parallaxScaleInput.setValue(material.materialInfo.parallaxScale)
            run {
                val parallaxScaleSliderInput = object : SliderInput("Parallax Scale", WebSlider.HORIZONTAL, 0, 100, (material.materialInfo.parallaxScale * 100).toInt()) {

                    override fun onValueChange(value: Int, delta: Int) {
                        parallaxScaleInput.setValue(value.toFloat() / 100f)
                        engine.scene.materialManager.changeMaterial(material.materialInfo.copyXXX(parallaxScale = value.toFloat() / 100f))
                        engine.eventBus.post(MaterialChangedEvent(material))
                    }
                }

                val groupPanelParallaxScale = GroupPanel(4, WebLabel("Parallax Scale"), parallaxScaleInput, parallaxScaleSliderInput)
                webComponentPanel.addElement(groupPanelParallaxScale)
            }
        }
        run {
            val parallaxBiasInput = object : LimitedWebFormattedTextField(0f, 1f) {
                override fun onChange(currentValue: Float) {
                    engine.scene.materialManager.changeMaterial(material.materialInfo.copyXXX(parallaxBias = currentValue))
                    engine.eventBus.post(MaterialChangedEvent(material))
                }
            }
            parallaxBiasInput.setValue(material.materialInfo.parallaxBias)
            run {
                val parallaxBiasSliderInput = object : SliderInput("Parallax Bias", WebSlider.HORIZONTAL, 0, 100, (material.materialInfo.parallaxScale * 100).toInt()) {

                    override fun onValueChange(value: Int, delta: Int) {
                        parallaxBiasInput.setValue(value.toFloat() / 100f)
                        engine.scene.materialManager.changeMaterial(material.materialInfo.copyXXX(parallaxBias = value.toFloat() / 100f))
                        engine.eventBus.post(MaterialChangedEvent(material))
                    }
                }

                val groupPanelParallaxBias = GroupPanel(4, WebLabel("Parallax Bias"), parallaxBiasInput, parallaxBiasSliderInput)
                webComponentPanel.addElement(groupPanelParallaxBias)
            }
        }

        run {
            val values = EnumSet.allOf(SimpleMaterial.MaterialType::class.java).toTypedArray()
            val materialTypeInput = WebComboBox(values, values.indexOf(material.materialInfo.materialType))
            materialTypeInput.addActionListener { e ->
                val selected = materialTypeInput.selectedItem as SimpleMaterial.MaterialType
                engine.scene.materialManager.changeMaterial(material.materialInfo.copyXXX(materialType = selected))
                engine.eventBus.post(MaterialChangedEvent(material))
            }
            val materialTypePanel = GroupPanel(4, WebLabel("Material Type"), materialTypeInput)
            webComponentPanel.addElement(materialTypePanel)
        }
        run {
            val values = EnumSet.allOf(SimpleMaterial.TransparencyType::class.java).toTypedArray()
            val transparencyTypeInput = WebComboBox(values, values.indexOf(material.materialInfo.transparencyType))
            transparencyTypeInput.addActionListener { e ->
                val selected = transparencyTypeInput.selectedItem as SimpleMaterial.TransparencyType
                engine.scene.materialManager.changeMaterial(material.materialInfo.copyXXX(transparencyType = selected))
                engine.eventBus.post(MaterialChangedEvent(material))
            }
            val materialTypePanel = GroupPanel(4, WebLabel("Transparency Type"), transparencyTypeInput)
            webComponentPanel.addElement(materialTypePanel)
        }
        run {
            val values = EnumSet.allOf(SimpleMaterial.ENVIRONMENTMAP_TYPE::class.java).toTypedArray()
            val environmentMapInput = WebComboBox(values, values.indexOf(material.materialInfo.environmentMapType))
            environmentMapInput.addActionListener { e ->
                val selected = environmentMapInput.selectedItem as SimpleMaterial.ENVIRONMENTMAP_TYPE
                engine.scene.materialManager.changeMaterial(material.materialInfo.copyXXX(environmentMapType = selected))
            }
            val groupPanelEnvironmentMapType = GroupPanel(4, WebLabel("Environment map type"), environmentMapInput)
            webComponentPanel.addElement(groupPanelEnvironmentMapType)
        }

        panels.add(webComponentPanel)
    }

    private fun copyShaderIfNotPresent(chosenFile: File, shaderFileInWorkDir: File) {
        if (!shaderFileInWorkDir.exists()) {
            try {
                FileUtils.copyFile(chosenFile, shaderFileInWorkDir)
            } catch (e: IOException) {
                e.printStackTrace()
            }

        }
    }

    private fun addMaterialInitCommand(material: SimpleMaterial) {
        init(material)
        val result = engine.gpuContext.calculate {
            InitMaterialCommand(material, engine.scene.materialManager).execute()
        }
        if (result.isSuccessful) {
            showNotification(NotificationIcon.plus, "SimpleMaterial changed")

            init(result.material)
            engine.eventBus.post(MaterialChangedEvent(material))
        } else {
            showNotification(NotificationIcon.error, "Not able to change materials")
        }
    }

    private fun showNotification(icon: NotificationIcon, text: String) {
        val notificationPopup = WebNotificationPopup()
        notificationPopup.setIcon(icon)
        notificationPopup.displayTime = 2000
        notificationPopup.content = WebLabel(text)
        NotificationManager.showNotification(notificationPopup)
    }
}
